src/main/java/org/myrobotlab/service/Mqtt.java

```diff
@@ -14,231 +18,390 @@
+  /**
+   * an instance connecting will send a connect message to this topic
    */
-  transient MemoryPersistence persistence;
+  protected String onConnectTopic = "mrl/onConnect";
 
-  MqttConnectOptions conOpt = new MqttConnectOptions();
+  /**
+   * an instance disconnecting will send a disconnect message to this topic
+   */
+  protected String onDisconnectTopic = "mrl/onDisconnect";
 
-  boolean cleanSession = true; // Non durable subscriptions
-  transient MqttAsyncClient client;
-  String clientId = String.format("%s@%s", getName(), getId());
-  boolean isConnected = false;
-  String topic = String.format("myrobotlab/%s", getId());
-  int port = 1883;
-  int qos = 2;
+  protected String password = null;
 
   /**
-   * not sure what this is supposed to be - perhaps a back-end id to identify
-   * the transaction on the listener
+   * for persistence of message if not sent immediately types include memory &
+   * file persistence
    */
-  String userContext = "context";
+  protected transient MemoryPersistence persistence;
 
-  Set<String> subscriptions = new HashSet<String>();
+  /**
+   * <pre>
+   * 0 - at most once (fastest most unreliable) 
+   * 1 - at least once (fast - guaranteed at least 1 - possible dupes - most common)
+   * 2 - exactly once (slowest - one and only one)
+   * </pre>
+   */
+  protected int qos = 1;
+
+  boolean retain = false;
 
   /**
-   * all incoming mqtt msgs
+   * reference to runtime route table
    */
-  String inTopic;
+  protected transient RouteTable routeTable = null;
+
+  protected transient Runtime runtime = null;
+
+  protected int rxCount;
+
+  protected String sslCaFilePath = null;
+
+  protected String sslCertFilePath = null;
+
+  protected String sslKeyFilePath = null;
+
+  protected String sslPassword = null;
 
   /**
-   * all outgoing mqtt msgs
+   * registry of merged instances of the form
+   * 
+   * <pre>
+   * { 
+   *   "{fullname}": "{type}",
+   *    "web01@admin":"WebGui",
+   *    ...   *
+   * </pre>
    */
-  String outTopic;
+  // protected String gatewayTopic = "mrl/gateway";
+
+  protected Set<String> subscriptions = new HashSet<String>();
 
-  // String url = "tcp://iot.eclipse.org:1883";
+  protected int txCount;
 
   /**
    * Two types of connection are supported tcp:// for a TCP connection and
    * ssl:// , which is weird, that its not mqtt:// and mqtts:// :P
+   * tcp://broker.hivemq.com:1883 is valid
    */
-  String url = "tcp://broker.hivemq.com:1883";
+  protected String url = "mqtt://localhost:1883";
 
   /**
-   * json msg codec
+   * not sure what this is supposed to be - perhaps a back-end id to identify
+   * the transaction on the listener
    */
+  protected String userContext = "context";
 
-  boolean autoSubscribe = true;
-  String userName = null;
-  char[] password = null;
-  boolean autoReconnect = true;
+  protected String username = null;
 
   public Mqtt(String n, String id) {
     super(n, id);
     try {

```

Comment:
could this stuff be done in a startService instead?  I'd like to move away from initialization code in the constructor.. and moving that code into startService..   this keeps the instantiation of objects lightweight.

Reply:
hmmm .. dunno, these are primitives btw .... not really very "heavy weight"